### PR TITLE
DIG-85 | Contextly.

### PR DESCRIPTION
**Pull request by dj1999**
WIP

### CODE REVIEW

File: config/sync/block.block.contextlywidgetblock_2.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+uuid: e79a896f-3682-40ef-85c2-c93db773d60a
+langcode: en
+status: true
+dependencies:
+  module:
+    - contextly
+    - diginomica
+  theme:
+    - diginomica_theme
+id: contextlywidgetblock_2
```
Comment (szatom) [inline config/sync/block.block.contextlywidgetblock_2.yml]: pls rename it, e.g. contextly_siderial

File: config/sync/block.block.contextlywidgetblock.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+uuid: c10067fd-62b8-4ba3-b933-d28a7d6aea25
+langcode: en
+status: true
+dependencies:
+  module:
+    - contextly
+    - diginomica
+  theme:
+    - diginomica_theme
+id: contextlywidgetblock
```
Comment (szatom) [inline config/sync/block.block.contextlywidgetblock.yml]: pls rename it, e.g. contextly_module

File: web/modules/custom/contextly/composer.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "drupal/contextly",
+  "type": "drupal-module",
+  "description": "Contextly module",
+  "keywords": ["Drupal"],
+  "license": "GPL-2.0+",
+  "homepage": "https://www.drupal.org/project/contextly",
```
Comment (szatom) [inline web/modules/custom/contextly/composer.json]: remove json file or add as custom repo

File: web/modules/custom/contextly/contextly.info.yml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+name: 'Contextly'
+type: module
+description: 'Contextly module'
```
Comment (szatom) [inline web/modules/custom/contextly/contextly.info.yml]: Add more information, e.g. like in D7 version:
Contextly service integration.

File: web/modules/custom/contextly/contextly.links.menu.yml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+contextly.contextly_admin_form:
+  title: 'Contextly Administration'
```
Comment (szatom) [inline web/modules/custom/contextly/contextly.links.menu.yml]: I think 'Contextly' is enough.

File: web/modules/custom/contextly/contextly.module [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    $page['#attached']['html_head'][] = [$contextly, 'contextly-page'];
+  }
+}
+
+/**
+ * Implements hook_preprocess_page().
+ *
+ * Add contextly libraries to page.
+ */
+function contextly_preprocess_page(array &$variables) {
```
Comment (szatom) [inline web/modules/custom/contextly/contextly.module]: Do we need for all pages? Not only for nodes/articles?

File: web/modules/custom/contextly/contextly.libraries.yml [hunk 7]
```diff
@@ -0,0 +1,12 @@
+contextly.head:
+  header: TRUE
+  js:
+    js/contextlyHeadMin.js: {attributes: {assync: TRUE}}
```
Comment (szatom) [inline web/modules/custom/contextly/contextly.libraries.yml]: minified

File: web/modules/custom/contextly/contextly.libraries.yml [hunk 8]
```diff
@@ -0,0 +1,12 @@
+contextly.head:
+  header: TRUE
```
Comment (szatom) [inline web/modules/custom/contextly/contextly.libraries.yml]: version?

File: web/modules/custom/contextly/contextly.libraries.yml [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+  js:
+    js/contextlyHeadMin.js: {attributes: {assync: TRUE}}
+  dependencies:
+    - core/jquery
+
+contextly.ready:
+  js:
```
Comment (szatom) [inline web/modules/custom/contextly/contextly.libraries.yml]: version?

File: web/modules/custom/contextly/js/contextlyReady.js [hunk 10]
```diff
@@ -0,0 +1 @@
+Contextly.ready('widgets');
```
Comment (szatom) [inline web/modules/custom/contextly/js/contextlyReady.js]: add new line

File: web/modules/custom/contextly/src/ContextlyWidgets.php [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+          '#tag' => 'div',
+          '#attributes' => [
+            'class' => [
+              'ctx-channel-container',
+              'ctx-clearfix',
+            ],
+          ],
+        ],
+      ],
+      'ctx_subscribe_container' => [
```
Comment (szatom) [inline web/modules/custom/contextly/src/ContextlyWidgets.php]: Duplicated 'ctx_subscribe_container' from line #100

File: web/modules/custom/contextly/src/MetaGeneratorInterface.php [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+
+/**
+ * Interface MetaGeneratorInterface.
+ */
+interface MetaGeneratorInterface {
+
+  /**
+   * Return the entity which generating metagata.
+   *
+   * @return Drupal\Core\Entity\EntityInterface
```
Comment (szatom) [inline web/modules/custom/contextly/src/MetaGeneratorInterface.php]: Should be (\ at the beginning)
\Drupal\Core\Entity\EntityInterface

File: web/modules/custom/contextly/src/Plugin/Block/ContextlyWidgetBlock.php [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+   *
+   * @var \Drupal\diginomica\ContainerInterface
+   *   The service container.
+   */
+  protected $container;
+
+  /**
+   * The Drupal\contextly\ContextlyWidgetsInterface definition.
+   *
+   * @var Drupal\contextly\ContextlyWidgetsInterface
```
Comment (szatom) [inline web/modules/custom/contextly/src/Plugin/Block/ContextlyWidgetBlock.php]: Should be (\ at the beginning)
\Drupal\contextly\ContextlyWidgetsInterface

File: web/modules/custom/contextly/src/Plugin/Block/ContextlyWidgetBlock.php [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+  /**
+   * Constructs a new ContextlyWidgetBlock object.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin_id for the plugin instance.
+   * @param string $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\diginomica\ContainerInterface $container
```
Comment (szatom) [inline web/modules/custom/contextly/src/Plugin/Block/ContextlyWidgetBlock.php]: I think it should be
Symfony\Component\DependencyInjection\ContainerInterface

File: web/modules/custom/contextly/src/Plugin/Block/ContextlyWidgetBlock.php [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+ *  id = "contextly_widget_block",
+ *  admin_label = @Translation("Contextly widget block"),
+ * )
+ */
+class ContextlyWidgetBlock extends BlockBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * The Drupal\diginomica\ContainerInterface definition.
+   *
+   * @var \Drupal\diginomica\ContainerInterface
```
Comment (szatom) [inline web/modules/custom/contextly/src/Plugin/Block/ContextlyWidgetBlock.php]: I think it should be
Symfony\Component\DependencyInjection\ContainerInterface

File: web/modules/custom/contextly/src/Form/ContextlyAdminForm.php [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+ * Class ContextlyAdminForm.
+ */
+class ContextlyAdminForm extends ConfigFormBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getEditableConfigNames() {
+    return [
+      'contextly.contextlyadmin',
```
Comment (szatom) [inline web/modules/custom/contextly/src/Form/ContextlyAdminForm.php]: I would prefer: contextly.settings

File: web/modules/custom/contextly/src/Form/ContextlyAdminForm.php [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+   */
+  public function getFormId() {
+    return 'contextly_admin_form';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    $config = $this->config('contextly.contextlyadmin');
```
Comment (szatom) [inline web/modules/custom/contextly/src/Form/ContextlyAdminForm.php]: I would prefer: contextly.settings

File: web/modules/custom/contextly/src/Form/ContextlyAdminForm.php [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+    return parent::buildForm($form, $form_state);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    parent::submitForm($form, $form_state);
+
+    $this->config('contextly.contextlyadmin')
```
Comment (szatom) [inline web/modules/custom/contextly/src/Form/ContextlyAdminForm.php]: I would prefer: contextly.settings

File: web/modules/custom/contextly/src/MetaGenerator.php [hunk 19]
```diff
@@ -0,0 +1,274 @@
...
+      }
+      else {
+        $function = $definition;
+        $class = $this;
+      }
+
+      call_user_func([$class, $function], $meta_key);
+    }
+
+    $this->container->get('module_handler')->alter('contextly_meta_tags', $this->metaTags, $this);
```
Comment (szatom) [inline web/modules/custom/contextly/src/MetaGenerator.php]: Custom hook definition, pls add documentation + example in
contextly.api.php

File: web/modules/custom/contextly/contextly.module (hunk 6 above)
Comment (szatom) [reply web/modules/custom/contextly/contextly.module]: Check the route 'entity.node.canonical'

File: web/modules/custom/contextly/contextly.module [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+  }
+}
+
+/**
+ * Implements hook_page_attachments().
+ *
+ * Add contextly meta tags to html head.
+ */
+function contextly_page_attachments(array &$page) {
+  if ($node = \Drupal::routeMatch()->getParameter('node')) {
```
Comment (szatom) [inline web/modules/custom/contextly/contextly.module]: Check the route 'entity.node.canonical'

File: web/modules/custom/contextly/contextly.module [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+ * @file
+ * Contains contextly.module.
+ */
+
+use Drupal\Core\Routing\RouteMatchInterface;
+
+/**
+ * Implements hook_help().
+ */
+function contextly_help(
```
Comment (szatom) [inline web/modules/custom/contextly/contextly.module]: empty implemetation.

Delete or add useful information.

File: web/modules/custom/contextly/src/MetaGenerator.php [hunk 22]
```diff
@@ -0,0 +1,276 @@
...
+
+    $this->metaTags[$key] = $url;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getApiData(): ? array {
+    $api_key = $this->container->get('config.factory')
+      ->get('contextly.contextlyadmin')->get('api_key');
```
Comment (szatom) [inline web/modules/custom/contextly/src/MetaGenerator.php]: settings

File: web/modules/custom/diginomica/diginomica.module [hunk 23]
```diff
@@ -364,3 +365,16 @@ function diginomica_field_widget_form_alter(&$element, FormStateInterface &$form
...
 }
+
+/**
+ * Implements hook_contextly_meta_tags_alter().
+ *
+ * Prepare image url from managed image field.
+ */
+function diginomica_contextly_meta_tags_alter(array &$meta_tags, MetaGeneratorInterface $meta_generator) {
+  $entity = $meta_generator->getEntity();
+  if (!empty($entity->field_featured_image->entity)) {
```
Comment (szatom) [inline web/modules/custom/diginomica/diginomica.module]: hasField() ?

Comment (dj1999) [reply web/modules/custom/diginomica/diginomica.module]: This is better performance. But if you want to using function than will change it.

**Pull request merged**